edx_arch_experiments/csp/middleware.py

```diff
@@ -0,0 +1,147 @@
+"""
+Middleware to add Content-Security-Policy and related headers.
+
+The ``content_security_policy_middleware`` middleware function can add
+``Content-Security-Policy``, ``Content-Security-Policy-Report-Only``, and
+``Reporting-Endpoints`` HTTP response headers. This functionality is
+configured by Django settings but is also gated behind a Waffle flag,
+``cms.experimental.content-security-policy``.
+"""
+import re
+import warnings
+
+from django.conf import settings
+from django.core.exceptions import MiddlewareNotUsed
+from edx_toggles.toggles import WaffleFlag
+
+
+def _load_headers() -> dict:
+    """
+    Return a dict of headers to append to every response, based on settings.
+    """
+    # .. setting_name: CSP_ENFORCE
+    # .. setting_default: None
+    # .. setting_description: Content-Security-Policy header to attach to all responses.
+    #   This should include everything but the ``report-to`` or ``report-uri`` clauses; those
+    #   will be appended automatically according to the ``CSP_REPORTING_NAME`` and
+    #   ``CSP_REPORTING_URI`` settings. Newlines are permitted and will be replaced with spaces.
+    #   A trailing `;` is also permitted.
+    # .. setting_warning: Setting the CSP header to too strict a value can cause your pages to
+    #   break. It is strongly recommended that deployers start by using ``CSP_REPORT_ONLY`` (along
+    #   with the reporting settings) and only move or copy the policies into ``CSP_ENFORCE`` after
+    #   confirming that the received CSP reports only represent false positives. (The report-only
+    #   and enforcement headers may be used at the same time.)
+    enforce_policies = getattr(settings, 'CSP_ENFORCE', None)
+
+    # .. setting_name: CSP_REPORT_ONLY
+    # .. setting_default: None
+    # .. setting_description: Content-Security-Policy-Report-Only header to attach to
+    #   all responses. See ``CSP_ENFORCE`` for details.
+    report_policies = getattr(settings, 'CSP_REPORT_ONLY', None)
+
+    # .. setting_name: CSP_REPORTING_URI
+    # .. setting_default: None
+    # .. setting_description: Content-Security-Policy-Report-Only header to attach to
+    #   all responses. See ``CSP_ENFORCE`` for details.
+    reporting_uri = getattr(settings, 'CSP_REPORTING_URI', None)
+
+    # .. setting_name: CSP_REPORTING_NAME
+    # .. setting_default: None
+    # .. setting_description: Used for CSP Level 3 reporting. This sets the name to use in the
+    #   report-to CSP field and the Reporting-Endpoints header. This should generally be an
+    #   alphanumeric string; other characters such as hyphen and underscore are also allowed.
+    #   See https://www.rfc-editor.org/rfc/rfc8941.html#section-3.3.4 for full grammar.
+    reporting_endpoint_name = getattr(settings, 'CSP_REPORTING_NAME', None)
+
+    # RFC 8941 Token is a RFC 7230 that must start with ALPHA or *, and can also contain / and :.
+    sfToken = "[a-zA-Z*][a-zA-Z0-9!#$%&'*+\\-.^_`|~\":/]*"
```

Comment:
Author's note: Not sure if I should bother with this arcane regex or if we should be stricter than the actual spec and just say "alphanumeric, hyphens, and underscore, but has to start with a letter" to make this check simpler.